jms/src/main/scala/akka/stream/alpakka/jms/JmsSourceStage.scala

```diff
@@ -5,80 +5,319 @@
 package akka.stream.alpakka.jms
 
 import java.util.concurrent.Semaphore
+import java.util.concurrent.atomic.AtomicBoolean
 import javax.jms._
 
-import akka.stream.stage.{GraphStage, GraphStageLogic, OutHandler, StageLogging}
-import akka.stream.{ActorAttributes, Attributes, Outlet, SourceShape}
+import akka.stream.stage._
+import akka.stream._
 
+import scala.annotation.tailrec
 import scala.collection.mutable
+import scala.concurrent.Future
+import scala.util.control.NonFatal
 import scala.util.{Failure, Success}
 
-final class JmsSourceStage(settings: JmsSourceSettings) extends GraphStage[SourceShape[Message]] {
+final class JmsSourceStage(settings: JmsSourceSettings)
+    extends GraphStageWithMaterializedValue[SourceShape[Message], KillSwitch] {
```

Comment:
Do we really need to materialize that killswitch? Can't the stage easily keep track of outstanding non-ack:ed `Message`s and keep going until those are done?

Reply:
Only if you use the proprietary `INDIVIDUAL_ACKNOWLEDGE`. Also, what is the formal way to stop the stream? Sticking an extra `KillSwitch` behind the source allows for messages to be fetched and not processed. This will cause re-delivery in the appropriate ack modes and transactional mode, and message loss in `AUTO_ACKNOWLEDGE`. Actually, IMO, any message consumer sources (Kafka included) should have a proper `KillSwitch` at materialization time. Kindly review the shutdown process in JmsSourceStage.